p2p/dandelion/src/txpool.rs

```diff
@@ -0,0 +1,545 @@
+    ///
+    /// This function does not add the tx to the backing pool.
+    async fn stem_tx(
+        &mut self,
+        tx: Tx,
+        tx_id: TxID,
+        from: Option<PID>,
+    ) -> Result<(), tower::BoxError> {
+        if let Some(peer) = &from {
+            self.stem_origins
+                .entry(tx_id.clone())
+                .or_default()
+                .insert(peer.clone());
+        }
+
+        let state = from
+            .map(|from| TxState::Stem { from })
+            .unwrap_or(TxState::Local);
+
+        let fut = self
+            .dandelion_router
+            .ready()
+            .await?
+            .call(DandelionRouteReq {
+                tx,
+                state: state.clone(),
+            });
+
+        self.routing_set
+            .spawn(fut.map(|res| (tx_id, res.map_err(|_| state))));
+        Ok(())
+    }
+
+    /// Stores the tx in the backing pool and fluffs the tx, removing the stem data for this tx.
+    async fn store_and_fluff_tx(&mut self, tx: Tx, tx_id: TxID) -> Result<(), tower::BoxError> {
+        // fluffs the tx first to prevent timing attacks where we could fluff at different average times
+        // depending on if the tx was in the stem pool already or not.
+        // Massively overkill but this is a minimal change.
+        self.fluff_tx(tx.clone(), tx_id.clone()).await?;
+
+        // Remove the tx from the maps used during the stem phase.
+        self.stem_origins.remove(&tx_id);
+
+        self.backing_pool
+            .ready()
+            .await?
+            .call(TxStoreRequest::Store(tx, tx_id, State::Fluff))
+            .await?;
+
+        // The key for this is *Not* the tx_id, it is given on insert, so just keep the timer in the
+        // map. These timers should be relatively short, so it shouldn't be a problem.
+        //self.embargo_timers.try_remove(&tx_id);
+
+        Ok(())
+    }
+
+    /// Fluffs a tx, does not add the tx to the tx pool.
+    async fn fluff_tx(&mut self, tx: Tx, tx_id: TxID) -> Result<(), tower::BoxError> {
+        let fut = self
+            .dandelion_router
+            .ready()
+            .await?
+            .call(DandelionRouteReq {
+                tx,
+                state: TxState::Fluff,
+            });
+
+        self.routing_set
+            .spawn(fut.map(|res| (tx_id, res.map_err(|_| TxState::Fluff))));
+        Ok(())
+    }
+
+    /// Function to handle an incoming [`DandelionPoolRequest::IncomingTx`].
+    async fn handle_incoming_tx(
+        &mut self,
+        tx: Tx,
+        tx_state: TxState<PID>,
+        tx_id: TxID,
+    ) -> Result<(), tower::BoxError> {
+        let TxStoreResponse::Contains(have_tx) = self
+            .backing_pool
+            .ready()
+            .await?
+            .call(TxStoreRequest::Contains(tx_id.clone()))
+            .await?
+        else {
+            panic!("Backing tx pool responded with wrong response for request.");
+        };
```

Comment:
Maybe a TODO for another time but `expected: Response::X(_), found: Response::Y(_)` would be helpful.